File: MDfiles/Data structures and Algorithms with Java Python and Javascript/JavaScript/testSearch.js
console.log('test search!');
//var arr= tArray();

function tArray(){
    var a=[]
    for (var i=0;i<10;i++){  
        a[i]=Math.floor((Math.random()*100)+1);
    }
     return a;
}
var arr = [41, 19, 35, 22, 84,39, 35, 64, 87, 76];
console.log(arr)

function seqSearch_(arr,data){

    for (var i=0;i<arr.length;i++){
        if (arr[i]==data){
            return true;
        }
    }
    return false;
}

function seqSearch(arr,data){
    for (var i=0;i<arr.length;i++){
        if (arr[i]==data){
            return i;
        }
    }
    return -1;
}
console.log(seqSearch(arr,41))

function findMin(arr){
    var min = arr[0];
    for (var i=1;i<arr.length;i++){
        if (arr[i]<min){
            min = arr[i];
        }
    }
    return min;
}
console.log("min =>"+findMin(arr));

function findMax(arr){
    var max = arr[0];
    for (var i=0;i<arr.length;i++){
         if (arr[i]>max){
             max = arr[i];
         }
    }
       return max;

}
console.log("max =>"+findMax(arr));

function binSearch(arr,data){
    var u = arr.length-1;
    var low =0;
    while (low<=u){
        var mid = Math.floor((u+low)/2);
        if (arr[mid]<data){
            low=mid+1;
        }else if (arr[mid]>data){
            u = mid-1;
        }else{
            return mid;
        }
    }
    return -1;
}

function insertSort(arr){
    var temp, i;
    for (var j=1;j<arr.length;j++){
        temp = arr[j];
        i = j;
        while (i>=0&&arr[i-1]>=temp){
            arr[i]=arr[i-1];
            --i;
        }
           arr[i] =temp;
    }
    return arr;
}

// search the matched data with last positon
function seqSearchM(arr,data){
    var max =0;
    for (var i=0;i<arr.length;i++){

        if (arr[i]==data&&i>(arr.length*0.2)){
            return i;
        }
    }
    return -1;
}
function seqSearchM__(arr,data){
    var p =[];
    for (var i=0;i<arr.length;i++){
        if (arr[i]==data){
            p.push(i);
        }
    }

    if (p.length==0){
        return -1;
    }else{
        var max =p[0];
        for (var i=0;i<p.length;i++){
            if (p[i]>max){
               max = p[i];
            }
        }
        return max;
    }
}

console.log('m=>'+seqSearchM__(arr,35));

File: MDfiles/Data structures and Algorithms with Java Python and Javascript/Python/testSearch.py
import pytest
from random import randint
#pytest testSearch.py::test_search -s


def test_search():
    nums = generatorNumbers()#[1,-1,9,10,100,2]
    print(nums)
    print('max=>',findMin(nums))

def findMax(nums):

    if len(nums)==1:
        return nums[0]
    op1 = nums[0]
    #print('op1=>',op1)
    op2 = findMax(nums[1:])
    #print('op2=>',op2)

    if op1>op2:
        return op1
    else:
        return op2

def findMin(nums):
    min = nums[0]
    for x in range(1,len(nums)):
        if nums[x]<min:
            min = nums[x]
    return min

def search(item,nums):

    for i in range(0,len(nums)):
        if nums[i]==item:
            return nums[i]


def bsort(arr):
    for i in range(len(arr)-1,0,-1):
        for j in range(i):
            if arr[j]<arr[j+1]:
                temp    = arr[j+1]
                arr[j+1]= arr[j]
                arr[j] = temp
    return arr

def binarySearch(item,nums):

    print('binarySearch!')
    n = len(nums)-1
    lowerIdex=0
    upperIdx=n

    while lowerIdex <= upperIdx:
        midPt = (lowerIdex+upperIdx)//2
        if item == nums[midPt]:
            return midPt
        if item >nums[midPt]:
            lowerIdex=midPt+1
        else:
            upperIdx=midPt-1

    if lowerIdex>upperIdx:
        return None

def findSameNums(nums):

    for i in range(0,len(nums)):
        for j in range(i+1,len(nums)):
            if nums[i]==nums[j]:
                print('local is',j,'=>',nums[j])

def generatorNumbers():
    nums=[]
    for _ in range(10):
        nums.append(randint(0, 100))
    return nums
